### ISSUE TITLE
Crash when rendering turtle upgrades

### ISSUE THREAD

**Issue by updogonline** [SquidDev-CC/CC-Tweaked#1626]
### Minecraft Version

1.20.1

### Version

cc-tweaked-1.20.1-forge-1.108.1.jar

### Details

 I am playing Prominence 2 on a server with some friends and when looking up stuff on jei the game crashes, ive narrowed it down to this mod, i get exit code -1, ive already tried increasing the memory allocated to the game
 
heres the crash report:

[crashlog.txt](https://github.com/cc-tweaked/CC-Tweaked/files/13324634/crashlog.txt)

this is the crash recreated using prominence 2 v1.4 i tried updating cc to the latest version and i still get the same error

help a gal out pls

**Comment by SquidDev** [comment]
Thanks for the report! This appears to be a crash caused by crafty turtle's upgrade model being `null`, which causes a crash later on. This is really odd - [the function which gets this model](https://github.com/cc-tweaked/CC-Tweaked/blob/76968f2f28435ddd8cf1efe5dc7330410e87cc09/projects/forge/src/client/java/dan200/computercraft/client/platform/ClientPlatformHelperImpl.java#L32-L34) shouldn't ever return `null`.

I suspect there's some optimisation mod which is causing issues here. It might be worth updating to the latest version of the whole pack (there are some changes to ModernFix which might be related), but will try to do some investigation to see what's going on.

**Edit:** Afraid I cannot reproduce this. Does this happen every time? If so, could you attach your full logs - just in case there's something helpful earlier on too.

**Comment by SquidDev** [comment]
It occurs to me that this is possibly the same issue as #1607. @NirioFaid, did you manage to work out what was going on?

**Comment by NirioFaid** [comment]
> It occurs to me that this is possibly the same issue as #1607. @NirioFaid, did you manage to work out what was going on?

Nope, and it still occurs, although sometimes it just works as intended.

The same MC 1.20.1, although it's the latest version of mod from CurseForge.

![image](https://github.com/cc-tweaked/CC-Tweaked/assets/55434483/c8a60701-ff92-4918-a2e3-a2d17c9c7b64)

**Comment by updogonline** [comment]
yep it happens every time, i asked a friend to try if they can reproduce it and it worked fine for them, im going to disable modernfix to see if that fixes it

i also tried updating to the latest pack version and the problem still stands

**Comment by updogonline** [comment]
update: disabling modernfix didnt work

attaching the crash report once again

[crash-2023-11-11_13.50.14-client.txt](https://github.com/cc-tweaked/CC-Tweaked/files/13326360/crash-2023-11-11_13.50.14-client.txt)

**Comment by SquidDev** [comment]
I tried deleting some models from the CC jar, just to see if that helped reproducing this bug. However, we just get the missing model, no crashes:

![A screenshot of the JEI sidebar in Minecraft, filtered to show all turtles. Some turtles have a missing model (the purple and black box).](https://github.com/cc-tweaked/CC-Tweaked/assets/4346137/c9cc93c3-ba71-46b2-932c-b11994a735ca)

I've done a quick search though any mixins in the pack, and I can't see anything obvious touching `ModelManager.getModel`.

Afraid I'm out of ideas here. It's probably worth reporting this [on the modpack's issue tracker too](https://github.com/LunaPixelStudios/Prominence/issues), just in case the pack author is able to track this down.

**Comment by SquidDev** [comment]
It appears people are still hitting this, and I'm still unable to reproduce.

I've [created a branch](https://github.com/cc-tweaked/CC-Tweaked/tree/hotfix/missing-model) with some additional debugging for this error.

Would you be able to [download the Forge jar from the CI build](https://github.com/cc-tweaked/CC-Tweaked/actions/runs/7386321607) and try to reproduce the bug? It should no longer crash, 
but will print some additional information to the console, which should help in debugging this.

**Comment by SquidDev** [comment]
#1667 also mentions that this bug is fixed on the latest version of the pack. Are other people able to confirm?

**Comment by raiomarco** [comment]
Found this crash on the 1.109.3, but with the block turtle...

I could help on reproducing it, i will try to isolate the problem and the mod to find what specifically breaks it

```
---- Minecraft Crash Report ----
// My bad.

Time: 2024-01-06 21:59:57
Description: Rendering Block Entity

java.lang.NoSuchMethodError: 'net.minecraft.resources.ResourceLocation dan200.computercraft.client.render.TurtleBlockEntityRenderer.getTurtleModel(dan200.computercraft.shared.computer.core.ComputerFamily, boolean)'
	at dan200.computercraft.client.render.TurtleBlockEntityRenderer.render(TurtleBlockEntityRenderer.java:590) ~[cc-tweaked-1.20.1-forge-1.109.3.jar%23615!/:1.109.3] {re:mixin,re:classloading,pl:mixin:APP:TurtleCosmetics.mixins.json:TurtleRenderingMixin,pl:mixin:APP:turtlematic.mixins.json:TurtleBlockEntityRendererMixin,pl:mixin:A}
	at dan200.computercraft.client.render.TurtleBlockEntityRenderer.m_6922_(TurtleBlockEntityRenderer.java:29) ~[cc-tweaked-1.20.1-forge-1.109.3.jar%23615!/:1.109.3] {re:mixin,re:classloading,pl:mixin:APP:TurtleCosmetics.mixins.json:TurtleRenderingMixin,pl:mixin:APP:turtlematic.mixins.json:TurtleBlockEntityRendererMixin,pl:mixin:A}
	at net.minecraft.client.renderer.blockentity.BlockEntityRenderDispatcher.m_112284_(BlockEntityRenderDispatcher.java:90) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,re:classloading,pl:accesstransformer:B,pl:mixin:APP:supplementaries.mixins.json:BlockEntityRendererDispatcherMixin,pl:mixin:APP:mixins.oculus.json:entity_render_context.MixinBlockEntityRenderDispatcher,pl:mixin:APP:flywheel.mixins.json:BlockEntityRenderDispatcherAccessor,pl:mixin:APP:embeddiumplus.mixin.json:entitydistance.MaxDistanceTileEntity,pl:mixin:APP:entityculling.mixins.json:BlockEntityRenderDispatcherMixin,pl:mixin:A}
	at net.minecraft.client.renderer.blockentity.BlockEntityRenderDispatcher.m_112297_(BlockEntityRenderDispatcher.java:79) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,re:classloading,pl:accesstransformer:B,pl:mixin:APP:supplementaries.mixins.json:BlockEntityRendererDispatcherMixin,pl:mixin:APP:mixins.oculus.json:entity_render_context.MixinBlockEntityRenderDispatcher,pl:mixin:APP:flywheel.mixins.json:BlockEntityRenderDispatcherAccessor,pl:mixin:APP:embeddiumplus.mixin.json:entitydistance.MaxDistanceTileEntity,pl:mixin:APP:entityculling.mixins.json:BlockEntityRenderDispatcherMixin,pl:mixin:A}
	at net.minecraft.client.renderer.blockentity.BlockEntityRenderDispatcher.m_112278_(BlockEntityRenderDispatcher.java:104) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,re:classloading,pl:accesstransformer:B,pl:mixin:APP:supplementaries.mixins.json:BlockEntityRendererDispatcherMixin,pl:mixin:APP:mixins.oculus.json:entity_render_context.MixinBlockEntityRenderDispatcher,pl:mixin:APP:flywheel.mixins.json:BlockEntityRenderDispatcherAccessor,pl:mixin:APP:embeddiumplus.mixin.json:entitydistance.MaxDistanceTileEntity,pl:mixin:APP:entityculling.mixins.json:BlockEntityRenderDispatcherMixin,pl:mixin:A}
	at net.minecraft.client.renderer.blockentity.BlockEntityRenderDispatcher.m_112267_(BlockEntityRenderDispatcher.java:79) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,re:classloading,pl:accesstransformer:B,pl:mixin:APP:supplementaries.mixins.json:BlockEntityRendererDispatcherMixin,pl:mixin:APP:mixins.oculus.json:entity_render_context.MixinBlockEntityRenderDispatcher,pl:mixin:APP:flywheel.mixins.json:BlockEntityRenderDispatcherAccessor,pl:mixin:APP:embeddiumplus.mixin.json:entitydistance.MaxDistanceTileEntity,pl:mixin:APP:entityculling.mixins.json:BlockEntityRenderDispatcherMixin,pl:mixin:A}
	at me.jellysquid.mods.sodium.client.render.SodiumWorldRenderer.renderBlockEntity(SodiumWorldRenderer.java:404) ~[embeddium-0.2.16+mc1.20.1.jar%23714!/:?] {re:mixin,re:classloading,pl:mixin:APP:mixins.oculus.compat.sodium.json:shadow_map.MixinSodiumWorldRenderer,pl:mixin:APP:mixins.oculus.compat.sodium.json:shadow_map.SodiumWorldRendererAccessor,pl:mixin:A}
	at me.jellysquid.mods.sodium.client.render.SodiumWorldRenderer.renderBlockEntities(SodiumWorldRenderer.java:335) ~[embeddium-0.2.16+mc1.20.1.jar%23714!/:?] {re:mixin,re:classloading,pl:mixin:APP:mixins.oculus.compat.sodium.json:shadow_map.MixinSodiumWorldRenderer,pl:mixin:APP:mixins.oculus.compat.sodium.json:shadow_map.SodiumWorldRendererAccessor,pl:mixin:A}
	at me.jellysquid.mods.sodium.client.render.SodiumWorldRenderer.renderBlockEntities(SodiumWorldRenderer.java:291) ~[embeddium-0.2.16+mc1.20.1.jar%23714!/:?] {re:mixin,re:classloading,pl:mixin:APP:mixins.oculus.compat.sodium.json:shadow_map.MixinSodiumWorldRenderer,pl:mixin:APP:mixins.oculus.compat.sodium.json:shadow_map.SodiumWorldRendererAccessor,pl:mixin:A}
	at net.minecraft.client.renderer.LevelRenderer.handler$zla000$onRenderBlockEntities(LevelRenderer.java:7273) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.renderer.LevelRenderer.m_109599_(LevelRenderer.java:1268) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.renderer.GameRenderer.m_109089_(GameRenderer.java:1126) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:advanced-debug-common.mixins.json:GameRendererMixin,pl:mixin:APP:supplementaries-common.mixins.json:GameRendererMixin,pl:mixin:APP:embeddium.mixins.json:features.gui.hooks.console.GameRendererMixin,pl:mixin:APP:mixins.oculus.json:GameRendererAccessor,pl:mixin:APP:mixins.oculus.json:MixinGameRenderer,pl:mixin:APP:mixins.oculus.json:MixinModelViewBobbing,pl:mixin:APP:immediatelyfast-common.mixins.json:core.compat.MixinGameRenderer,pl:mixin:APP:ad_astra-common.mixins.json:client.GameRendererMixin,pl:mixin:APP:embeddiumplus.mixin.json:darkness.MixinGameRenderer,pl:mixin:APP:mixins.satin.client.json:event.GameRendererMixin,pl:mixin:APP:forge-sfcr-common.mixins.json:GameRendererMixin,pl:mixin:APP:fastload.mixins.json:client.GameRendererMixin,pl:mixin:APP:ars_nouveau.mixins.json:GameRendererMixin,pl:mixin:APP:immersive_aircraft.mixins.json:GameRendererMixin,pl:mixin:APP:railways-common.mixins.json:conductor_possession.MixinGameRenderer,pl:mixin:APP:seamless_loading_screen-common.mixins.json:GameRendererMixin,pl:mixin:APP:sodium-extra.mixins.json:prevent_shaders.MixinGameRenderer,pl:mixin:APP:controllable.common.mixins.json:client.GameRendererMixin,pl:mixin:APP:controllable.mixins.json:client.ForgeGameRendererMixin,pl:mixin:APP:alexscaves.mixins.json:client.GameRendererMixin,pl:mixin:APP:create.mixins.json:accessor.GameRendererAccessor,pl:mixin:APP:create.mixins.json:client.GameRendererMixin,pl:mixin:APP:mixins.oculus.json:MixinGameRenderer_NightVisionCompat,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.renderer.GameRenderer.m_109093_(GameRenderer.java:909) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:advanced-debug-common.mixins.json:GameRendererMixin,pl:mixin:APP:supplementaries-common.mixins.json:GameRendererMixin,pl:mixin:APP:embeddium.mixins.json:features.gui.hooks.console.GameRendererMixin,pl:mixin:APP:mixins.oculus.json:GameRendererAccessor,pl:mixin:APP:mixins.oculus.json:MixinGameRenderer,pl:mixin:APP:mixins.oculus.json:MixinModelViewBobbing,pl:mixin:APP:immediatelyfast-common.mixins.json:core.compat.MixinGameRenderer,pl:mixin:APP:ad_astra-common.mixins.json:client.GameRendererMixin,pl:mixin:APP:embeddiumplus.mixin.json:darkness.MixinGameRenderer,pl:mixin:APP:mixins.satin.client.json:event.GameRendererMixin,pl:mixin:APP:forge-sfcr-common.mixins.json:GameRendererMixin,pl:mixin:APP:fastload.mixins.json:client.GameRendererMixin,pl:mixin:APP:ars_nouveau.mixins.json:GameRendererMixin,pl:mixin:APP:immersive_aircraft.mixins.json:GameRendererMixin,pl:mixin:APP:railways-common.mixins.json:conductor_possession.MixinGameRenderer,pl:mixin:APP:seamless_loading_screen-common.mixins.json:GameRendererMixin,pl:mixin:APP:sodium-extra.mixins.json:prevent_shaders.MixinGameRenderer,pl:mixin:APP:controllable.common.mixins.json:client.GameRendererMixin,pl:mixin:APP:controllable.mixins.json:client.ForgeGameRendererMixin,pl:mixin:APP:alexscaves.mixins.json:client.GameRendererMixin,pl:mixin:APP:create.mixins.json:accessor.GameRendererAccessor,pl:mixin:APP:create.mixins.json:client.GameRendererMixin,pl:mixin:APP:mixins.oculus.json:MixinGameRenderer_NightVisionCompat,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91383_(Minecraft.java:1146) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91374_(Minecraft.java:718) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at fudge.notenoughcrashes.mixinhandlers.InGameCatcher.handleClientCrash(InGameCatcher.java:37) ~[notenoughcrashes-4.4.6+1.20.1-forge.jar%23845!/:?] {re:mixin,re:classloading}
	at net.minecraft.client.Minecraft.modify$enm000$atTheEndOfFirstCatchBeforePrintingCrashReport(Minecraft.java:21511) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91374_(Minecraft.java:738) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at fudge.notenoughcrashes.mixinhandlers.InGameCatcher.handleClientCrash(InGameCatcher.java:37) ~[notenoughcrashes-4.4.6+1.20.1-forge.jar%23845!/:?] {re:mixin,re:classloading}
	at net.minecraft.client.Minecraft.modify$enm000$atTheEndOfFirstCatchBeforePrintingCrashReport(Minecraft.java:21511) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91374_(Minecraft.java:738) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at fudge.notenoughcrashes.mixinhandlers.InGameCatcher.handleClientCrash(InGameCatcher.java:37) ~[notenoughcrashes-4.4.6+1.20.1-forge.jar%23845!/:?] {re:mixin,re:classloading}
	at net.minecraft.client.Minecraft.modify$enm000$atTheEndOfFirstCatchBeforePrintingCrashReport(Minecraft.java:21511) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91374_(Minecraft.java:738) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at fudge.notenoughcrashes.mixinhandlers.InGameCatcher.handleClientCrash(InGameCatcher.java:37) ~[notenoughcrashes-4.4.6+1.20.1-forge.jar%23845!/:?] {re:mixin,re:classloading}
	at net.minecraft.client.Minecraft.modify$enm000$atTheEndOfFirstCatchBeforePrintingCrashReport(Minecraft.java:21511) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91374_(Minecraft.java:738) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at fudge.notenoughcrashes.mixinhandlers.InGameCatcher.handleClientCrash(InGameCatcher.java:37) ~[notenoughcrashes-4.4.6+1.20.1-forge.jar%23845!/:?] {re:mixin,re:classloading}
	at net.minecraft.client.Minecraft.modify$enm000$atTheEndOfFirstCatchBeforePrintingCrashReport(Minecraft.java:21511) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91374_(Minecraft.java:738) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at fudge.notenoughcrashes.mixinhandlers.InGameCatcher.handleClientCrash(InGameCatcher.java:37) ~[notenoughcrashes-4.4.6+1.20.1-forge.jar%23845!/:?] {re:mixin,re:classloading}
	at net.minecraft.client.Minecraft.modify$enm000$atTheEndOfFirstCatchBeforePrintingCrashReport(Minecraft.java:21511) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91374_(Minecraft.java:738) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at fudge.notenoughcrashes.mixinhandlers.InGameCatcher.handleClientCrash(InGameCatcher.java:37) ~[notenoughcrashes-4.4.6+1.20.1-forge.jar%23845!/:?] {re:mixin,re:classloading}
	at net.minecraft.client.Minecraft.modify$enm000$atTheEndOfFirstCatchBeforePrintingCrashReport(Minecraft.java:21511) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91374_(Minecraft.java:738) ~[client-1.20.1-20230612.114412-srg.jar%23958!/:?] {re:mixin,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick,xf:fml:xaeroworldmap:xaero_wm_minecraft_runtick_render_call,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.main.Main.main(Main.java:218) ~[1.20.1-1.20.1-47.2.0.jar:?] {re:mixin,pl:runtimedistcleaner:A,re:classloading,pl:mixin:APP:flywheel.mixins.json:ClientMainMixin,pl:mixin:APP:smoothboot.mixins.json:client.MainMixin,pl:mixin:APP:notenoughcrashes.forge.mixins.json:client.MixinMain,pl:mixin:A,pl:runtimedistcleaner:A}
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?] {}
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?] {}
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?] {}
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?] {re:mixin}
	at net.minecraftforge.fml.loading.targets.CommonLaunchHandler.runTarget(CommonLaunchHandler.java:111) ~[fmlloader-1.20.1-47.2.0.jar:?] {}
	at net.minecraftforge.fml.loading.targets.CommonLaunchHandler.clientService(CommonLaunchHandler.java:99) ~[fmlloader-1.20.1-47.2.0.jar:?] {}
	at net.minecraftforge.fml.loading.targets.CommonClientLaunchHandler.lambda$makeService$0(CommonClientLaunchHandler.java:25) ~[fmlloader-1.20.1-47.2.0.jar:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandlerDecorator.launch(LaunchServiceHandlerDecorator.java:30) ~[modlauncher-10.0.9.jar:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:53) ~[modlauncher-10.0.9.jar:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:71) ~[modlauncher-10.0.9.jar:?] {}
	at cpw.mods.modlauncher.Launcher.run(Launcher.java:108) ~[modlauncher-10.0.9.jar:?] {}
	at cpw.mods.modlauncher.Launcher.main(Launcher.java:78) ~[modlauncher-10.0.9.jar:?] {}
	at cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:26) ~[modlauncher-10.0.9.jar:?] {}
	at cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:23) ~[modlauncher-10.0.9.jar:?] {}
	at cpw.mods.bootstraplauncher.BootstrapLauncher.main(BootstrapLauncher.java:141) ~[bootstraplauncher-1.1.2.jar:?] {}
```
```
-- Block Entity Details --
Details:
	Name: computercraft:turtle_normal // dan200.computercraft.shared.turtle.blocks.TurtleBlockEntity
	Block: Block{computercraft:turtle_normal}[facing=west,waterlogged=false]
	Block location: World: (892,63,-1319), Section: (at 12,15,9 in 55,3,-83; chunk contains blocks 880,-64,-1328 to 895,319,-1313), Region: (1,-3; contains chunks 32,-96 to 63,-65, blocks 512,-64,-1536 to 1023,319,-1025)
	Block: Block{computercraft:turtle_normal}[facing=west,waterlogged=false]
	Block location: World: (892,63,-1319), Section: (at 12,15,9 in 55,3,-83; chunk contains blocks 880,-64,-1328 to 895,319,-1313), Region: (1,-3; contains chunks 32,-96 to 63,-65, blocks 512,-64,-1536 to 1023,319,-1025)
	Block Entity NBT: {ForgeCaps:{},Fuel:0,Items:[],On:0b,Slot:0}
```

**Comment by SquidDev** [comment]
@raiomarco That appears to be a separate issue caused by Turtle Cosmetics having a dodgy mixin. I've reported that at https://github.com/JheffersonMarques/TurtleCosmetics/issues/1.

Downgrading to 1.109.2 should resolved it until TurtleCosmtics have updated.

**Comment by laboriens** [comment]
for my part these applied energetics which crash cc:Tweaked

**Comment by CaelTheColher** [comment]
I'm experiencing the same crash on a different modpack
[crash-2024-01-14_23.52.39-client.txt](https://github.com/cc-tweaked/CC-Tweaked/files/13934424/crash-2024-01-14_23.52.39-client.txt)

Some comments:

- Funnily enough I happen to have the exact CPU model as OP, but I don't think its related to the crash
- The debug version crashes on launch for me
[crash-2024-01-15_00.06.07-client.txt](https://github.com/cc-tweaked/CC-Tweaked/files/13934446/crash-2024-01-15_00.06.07-client.txt)
- I also tried testing this issue without Sinytra Connector and while it didn't crash, the models were still bugged. Which makes me think this specific crash is an interaction between a mod making the turtle upgrade models be null and then Connector (or a related mod) causing it to crash

**Comment by SquidDev** [comment]
> I'm experiencing the same crash on a different modpack

Is this modpack publicly available anywhere? 

> The debug version crashes on launch for me

Thank you for testing, and sorry about that. It looks like the builds in CI aren't reobfuscated correctly - I'm not sure why, as the build process is the same everywhere.

Could you try [with this version instead](https://github.com/cc-tweaked/CC-Tweaked/files/13938403/cc-tweaked-1.20.1-forge-1.109.3-2e6dffab5e962fc2d8800f40df5673b8ae3a842f.zip)?

**Comment by CaelTheColher** [comment]
> Is this modpack publicly available anywhere?

Its just one a friend put together so unfortunately no, but I could send the mods folder to you in private if you want.

 >Could you try with this version instead?

This one works, but I only seem to get the specific log message when Connector is installed? Even though without Connector the models are still missing.

```
[Render thread/ERROR] [da.co.cl.pl.ClientPlatformHelperImpl/]: ModelManager.getModel("computercraft:block/turtle_crafting_table_right") returned null. This should never happen, and indicates the model registry is corrupt.

ModelManager has been transformed by the following transformers: re:mixin,pl:accesstransformer:B,pl:connector_pre_launch:A,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:rei_plugin_compatibilities:B,pl:mixin:A,pl:connector_pre_launch:A,pl:runtimedistcleaner:A.
The following models are null: .
```

And some updates: Apparently even with the exact same modpack I'm the only one in my friends group that is having issues, and yesterday I spent 6 hours trying to manually pinpoint the mod that is causing the missing models but couldn't, I think it might be more than one mod or even some external factor that might be contributing

**Comment by SquidDev** [comment]
Good shout about about Sinytra Connector - it looks like the crash was being caused by Fabric API conflicting with Forge, I've raised an issue about that.

As far as why the models are not loading, that's much odder. From what we saw in #1607, we're not even getting any missing model messages. Just to confirm, if you run `/give @p computercraft:turtle_normal{Overlay:"computercraft:block/turtle_rainbow_overlay"}`, does that turtle also have a missing model? I'm wondering if the extra model loading hooks just aren't being called at all or something odd.

**Comment by CaelTheColher** [comment]
That specific turtle does seem to work. The only ones being affected are the Crafty, Ender, Wireless and Noisy turtles just like #1607 shows.

**Comment by SquidDev** [comment]
Thank you for testing, that's really helpful! Though also very odd - that suggests some models are loaded fine, but not others. Does reloading resources (F3+T I believe) fix any of the models? Almost wondering if this is caused by a race condition in mod loading or something. 

**Edit:** To clarify, it's possible that `ModelEvent.RegisterAdditional` and `FMLClientSetupEvent` are dispatched in parallel rather than serially. This might mean that turtle upgrades aren't actually registered by the time the event is dispatched, so resource reloading would fix it.

I've added some logging and it is not clear to me quite how the different codepaths interact, will try to do some proper debugging tomorrow when I've more time.

**Comment by CaelTheColher** [comment]
>Does reloading resources (F3+T I believe) fix any of the models?

That indeed does fix all of the models!

My current theory is that the race condition happens due to my CPU (and therefore OP's since its the same model) isn't fast enough? And looking at the other known instances of this bug, all of them seem to be using laptops or weaker/older desktop cpus. This would also explain why this bug seems to be so elusive.

**Comment by SquidDev** [comment]
> It's possible that `ModelEvent.RegisterAdditional` and `FMLClientSetupEvent` are dispatched in parallel rather than serially.

This appears to be the case. FML has three loading stages:
 - `GATHER` (constructs mods and fires registry events)
 - `LOAD` (configs, the various setup events)
 - `COMPLETE` (IMC, `FMLLoadCompleteEvent`)

The `GATHER` phase is ]executed immediately when the game starts]((https://github.com/neoforged/NeoForge/blob/39661d944e8083890621494e8f1b0cd06dd63d23/src/main/java/net/neoforged/neoforge/client/loading/ClientModLoader.java#L57)). However, `LOAD` and `COMPLETE` are executed [inside a resource reload listener](https://github.com/neoforged/NeoForge/blob/39661d944e8083890621494e8f1b0cd06dd63d23/src/main/java/net/neoforged/neoforge/client/loading/ClientModLoader.java#L67-L71), which (I believe) means they're executed in parallel with resource loading.

I do find this a little surprising - I've posted a question on the NeoForge Discord to check my understanding of the code is correct.

If this is the case, I suspect we'll have to deprecate `registerTurtleUpgradeModeller` and handle registration inside an event instead. We should probably PR some documentation to [this page](https://docs.neoforged.net/docs/concepts/lifecycle) too to clarify this gotcha.

**Comment by laboriens** [comment]
the curseforge version no longer crashes since I updated the Forgified Fabric API mods, but the textures are not there

**Comment by SquidDev** [comment]
Yep, that's expected! The crash is caused by FFAPI, but the model loading issue _is_ a CC:T bug. That'll be fixed by the above PR.
